helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/RebalanceAlgorithm.java

```diff
@@ -19,26 +19,25 @@
  * under the License.
  */
 
+import org.apache.helix.HelixRebalanceException;
 import org.apache.helix.controller.rebalancer.waged.model.ClusterModel;
-import org.apache.helix.model.ResourceAssignment;
-
-import java.util.Map;
+import org.apache.helix.controller.rebalancer.waged.model.OptimalAssignment;
 
 /**
- * A generic rebalance algorithm interface for the WAGED rebalancer.
+ * A generic interface to generate the optimal assignment given the runtime cluster environment.
  *
- * @see <a href="Rebalance Algorithm">https://github.com/apache/helix/wiki/Design-Proposal---Weight-Aware-Globally-Even-Distribute-Rebalancer#rebalance-algorithm-adapter</a>
+ * <pre>
+ * @see <a href="https://github.com/apache/helix/wiki/
+ * Design-Proposal---Weight-Aware-Globally-Even-Distribute-Rebalancer
+ * #rebalance-algorithm-adapter">Rebalance Algorithm</a>
+ * </pre>
  */
 public interface RebalanceAlgorithm {
 
   /**
    * Rebalance the Helix resource partitions based on the input cluster model.
-   *
-   * @param clusterModel
-   * @param failureReasons Return the failures <ResourceName, <FailureReason, Count>> that happen during the rebalance calculation.
-   *                       If the map is null, no failure will be returned.
-   * @return A map of <ResourceName, ResourceAssignment>.
+   * @param clusterModel The run time cluster model that contains all necessary information
+   * @return An instance of {@link OptimalAssignment}
    */
-  Map<String, ResourceAssignment> rebalance(ClusterModel clusterModel,
-      Map<String, Map<HardConstraint, Integer>> failureReasons);
+  OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebalanceException;
```

Comment:
The HelixRebalanceException was designed for the rebalancer exception. That why we need the type there. Note that FAILED_TO_CALCULATE is already a rebalance error type there.
I'm not sure if it fits the algorithm implementation? I think you can have a different exception type if necessary.

Reply:
Another option is that we can just remove my try-catch blocks around all the algorithm.calculate() calls in the rebalancer code. Both works.

Please let me know your preference.

Reply:
FAILED_TO_CALCULATE is a little too general and it could also mean exception occurred in algorithm part. Adding another enum inside type is valid IMO, could you make the `FAILED_TO_CALCULATE` more specific, what calculation it should refer to?

Reply:
No, the problem is that HelixRebalanceException is for the rebalancer. So FAILED_TO_CALCULATE is the only thing we need to know. If you have more types to share, please implement an additional exception class.

Reply:
It makes sense if `FAILED_TO_CALCULATE` is used inside the algorithm implementation and taken care by rebalancer class.